### ISSUE TITLE
False positive unexpected `where`

### ISSUE THREAD

**Issue by neonaot** [intellij-rust/intellij-rust#9661]
## Environment

* **IntelliJ Rust plugin version:** 0.4.183.4982-223-nightly
* **Rust toolchain version:** 1.67.0-nightly (11ebe6512 2022-11-01) x86_64-apple-darwin
* **IDE name and version:** CLion 2022.3 EAP (CL-223.7255.15)
* **Operating system:** macOS 12.5.1
* **Macro expansion:** enabled
* **Additional experimental features:** org.rust.cargo.features.settings.gutter, org.rust.cargo.emulate.terminal, org.rust.macros.proc.function-like, org.rust.macros.proc

## Problem description
<img width="765" alt="image" src="https://user-images.githubusercontent.com/49211026/199477085-c591d930-08b9-4f65-a5cd-85af57a72b5b.png">

## Steps to reproduce
```rust
use std::fmt::Debug;
use std::marker::PhantomData;

trait Foo {
    type Gat<'a>: ?Sized where Self: 'a;
}

struct Bar<'a, T: Foo + 'a>(T::Gat<'a>);

struct Baz<T: ?Sized>(PhantomData<T>);

impl<T: ?Sized> Foo for Baz<T> {
    type Gat<'a> = T where Self: 'a; // '!', '+', '::', ';', <path parameters> or <type argument list> expected, got 'where'
}

fn main() {
    let x = Bar::<'_, Baz<()>>(());
    let y: &Bar<'_, Baz<dyn Debug>> = &x;
}
```

**Comment by Sherlock-Holo** [comment]
the rust realease notes gat example also got this problem

```rust
trait LendingIterator {
    type Item<'a>
    where
        Self: 'a;
}

pub struct WindowsMut<'x, T> {
    slice: &'x [T],
}

impl<'x, T> LendingIterator for WindowsMut<'x, T> {
    type Item<'a> = &'a mut [T] where Self: 'a;
}
```

![image](https://user-images.githubusercontent.com/10096425/203507388-348482b3-7e35-474f-8b0e-e9a944258e2a.png)
